pkg/controllers/clusterid.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright © 2024 MicroShift Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package controllers
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	"github.com/openshift/microshift/pkg/config"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+)
+
+type ClusterID struct {
+	cfg *config.Config
+}
+
+func NewClusterID(cfg *config.Config) *ClusterID {
+	s := &ClusterID{}
+	s.cfg = cfg
+	return s
+}
+
+func (s *ClusterID) Name() string { return "cluster-id-manager" }
+func (s *ClusterID) Dependencies() []string {
+	return []string{"kube-apiserver"}
+}
+
+func (s *ClusterID) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
+	defer close(ready)
+
+	// Read the 'kube-system' namespace attributes
+	restConfig, err := clientcmd.BuildConfigFromFlags("", s.cfg.KubeConfigPath(config.KubeAdmin))
+	if err != nil {
+		panic(err)
+	}
+	coreClient := clientv1.NewForConfigOrDie(rest.AddUserAgent(restConfig, "core-agent"))
+	namespace, err := coreClient.Namespaces().Get(ctx, "kube-system", metav1.GetOptions{})
+	if err != nil {
+		panic(err)
+	}
+
+	// Use the 'kube-system' namespace metadata UID as the MicroShift Cluster ID
+	clusterID := string(namespace.ObjectMeta.UID)
+	// Write <config.DataDir>/cluster-id file if it does not already exist
+	initClusterIDFile(clusterID)
+	// Log the cluster ID
+	klog.Infof("MicroShift Cluster ID: %v", clusterID)
+
+	return ctx.Err()
+}
+
+func initClusterIDFile(clusterID string) {
+	// The location of the cluster ID file
+	fileName := filepath.Join(config.DataDir, "cluster-id")
+
+	// Do not create the cluster ID file if it already exists
```

Comment:
Have we decided that getting this out of sync is non-issue? For example, what if user writes something to that file.

Reply:
I think it's not an issue. The file is created with `0400` permissions. Users can tamper with any database file potentially and get wierd results.

Do you think I should add a check between the namespace value and the file value? It's easy to detect such an inconsistency and abort.

Reply:
I thought we would overwrite the file on every startup to ensure the value is as close to up to date as possible.

Reply:
I would rather create the file once and not touch it. If it's tampered by the user, we could log an error message and exit. 
Otherwise, it may be a bit of a problem to explain this file purpose in the user docs, i.e. we have this read-only file here, but it is overwritten on every startup...

Reply:
So, I'm leaning to the stricter implementation, but I'm open to suggestions either way. 
WDYT?

Reply:
The only reason we're writing the file is to have a way to get the cluster ID when MicroShift is down and it can't be read from the database. I don't think it's good practice to refuse to start because the user has taken an action we can easily circumvent. This is a good example of how we can make the software more resilient by making it flexible.

Reply:
You could draw a parallel to a status field in a kube API. If this value was in the database and we wanted to prevent it from being overwritten we would have a controller constantly watch it and reset the value if someone changed it.

Reply:
Added a warning and rewrite in case of tampering with the cluster ID.
Also added a test checking this scenario.